src/charm.py

```diff
@@ -238,9 +242,17 @@ def _on_collect_unit_status(self, event: CollectStatusEvent):
             event.add_status(to_status(status))
 
     def _on_config_changed(self, _):
+        # If migration date isn't present then this is a fresh loki installation, hence set the v12 schema date to today
+        if not self.v12_migration_date:
+            self.v12_migration_date = (datetime.date.today()).strftime("%Y-%m-%d")
         self._configure()
 
     def _on_upgrade_charm(self, _):
```

Comment:
In the [upgrade sequence](https://discourse.charmhub.io/t/event-upgrade-charm/6485), upgrade-charm always fires before config-changed. However, afaik, there are no guarantees about `can_connect` at this point. This means that the following scenario would be possible:

1. upgrade a charm that already uses v12/tsdb.
2. (stop)
3. upgrade-charm fires before pebble-ready, so `self.v12_migration_date` is `""`. Hook changes it to tomorrow. 
    -  `_configure` is called.
    - (meanwhile, the container becomes available and the next `can_connect` is true)
    - a config file with "tsdb from tomorrow" is written to disk
4. config-changed fires, doesn't modify `self.v12_migration_date`
5. loki uses boltdb on a database that is already tsdb.

If you agree with the above, then one possible solution is:
1. Move the logic under `_configure`, after the can_connect guard.
2. If the backup exists, but our file doesn't, then it's an upgrade. If neither exist then it's a fresh install.

Wdyt?